### ISSUE TITLE
[Feature Request]: Debugger - Implement `undo` for `Add Function` `Rename Function` `Remove Function`

### ISSUE THREAD

**Issue by Risae** [PCSX2/pcsx2#7781]
### Description

It would be great if the Qt Debugger is able to have some kind of "undo", which can revert changes made by `Add Function` `Rename Function` `Remove Function`.

### Reason

It would be great if the Qt Debugger is able to have some kind of "undo", which can revert changes made by `Add Function` `Rename Function` `Remove Function`.

### Examples

It would be great if the Qt Debugger is able to have some kind of "undo", which can revert changes made by `Add Function` `Rename Function` `Remove Function`.

**Comment by buruto** [comment]
> It would be great if the Qt Debugger is able to have some kind of "undo", which can revert changes made by `Add Function`, `Rename Function` & `Remove Function`.

And what about `Assemble new instruction(s)` and `NOP Instruction(s)`? They should not have any `undo` to disable patches that the user did?

**Comment by Risae** [comment]
> > It would be great if the Qt Debugger is able to have some kind of "undo", which can revert changes made by `Add Function`, `Rename Function` & `Remove Function`.
> 
> And what about `Assemble new instruction(s)` and `NOP Instruction(s)`? They should not have any `undo` to disable game patches that the user did?
> 
> Also why your feature request is not written in the Description, Reason and Examples format?
> 
> How can we know how much your feature is useful so that it worthes development time?
> 
> Though your feature can be implemented in a day according to it's complexity but I don't know any use case for this feature.

The ability to have these features will improve the Debugger user experience, since often you change things in the Debugger and want to revert to a previous state. Right now you have to do this manually and make notes of the original code that you are changing.

**Comment by DIO6150** [comment]
Hello, this issue was recommended as a good first issue. Can it be assigned to me please ?

**Comment by GovanifY** [comment]
@DIO6150 Hi, we don't really assign issues to persons outside of our team, moreso issue assignment is more to indicate to people we are working on something and who to contact, on top of a personal tracking tool.

If you're new to FOSS I just want you to know that you can just do the change, open a PR, say this fixes this issue and if your change gets merged, then this issue will automatically close :)

**Comment by DIO6150** [comment]
@GovanifY Yes i'm a total newbie but I understand, thank you for your reply !

**Comment by chaoticgd** [comment]
I'm the one that originally wrote a lot of this code, some here's some ideas for implementing this:
- The function names are stored in the [symbol database](https://github.com/PCSX2/pcsx2/tree/master/3rdparty/ccc), which is accessed via the [symbol guardian](https://github.com/PCSX2/pcsx2/blob/master/pcsx2/DebugTools/SymbolGuardian.cpp).
- To implement undo/redo for the symbols, you should thus control access to the symbol database so that everywhere you write to it, you instead push a command object onto an undo/redo stack that supports applying and undoing the operation (or possibly clear the undo/redo stack for large operations). See how Qt handles this [here](https://doc.qt.io/qt-6/qundo.html), although the symbol databases are modified by more code than just the debugger, so you should consider how you'll keep these components in sync.
- Symbols reference each other using handles. Thus, if you want the database to remain consistent after an undo operation, you should make it so the symbols being recreated have their original handles, instead of just creating entirely new symbols. This will probably require modifying the code for the symbol database itself.

I'm not sure if I would say this is a good first issue, undo/redo can be kind of tricky to implement in large existing codebases (not that the debugger is that large), but if you want to do it go ahead.
